ktoolbox/common.py

```diff
@@ -1533,3 +1534,59 @@ def get_poll_fd(self) -> int:
                 if self._event.is_set():
                     os.write(self._write_fd, b"1")
             return self._read_fd
+
+
+class FutureThread(threading.Thread, typing.Generic[T1]):
+    def __init__(
+        self,
+        func: typing.Callable[["FutureThread[T1]"], T1],
+        *,
+        start: bool = False,
+        cancellable: Optional[Cancellable] = None,
+    ):
+        super().__init__()
+        self._lock = threading.Lock()
+        self._func = func
+        self.future: Future[T1] = Future()
+        self._cancellable = cancellable
+        self._cancellable_is_cancelled = False
+        self._user_data: Any = None
+        if start:
+            self.start()
+
+    @property
+    def cancellable(self) -> Cancellable:
+        with self._lock:
+            if self._cancellable is None:
+                self._cancellable = Cancellable()
+                if self._cancellable_is_cancelled:
```

Comment:
Why not check self._cancellable.is_cancelled?

Reply:
the `FutureThread`'s `self._cancellable` is only created if a caller requests it. This probably could be simplified by letting `FutureThread.__init__()` always create a cancellable, but the way it is, it's optimized to only create it once it's used.

Just look at the 3 uses of `_cancellable_is_cancelled`. As we create the cancellable only when requested, it is necessary to remember if the cancellable was cancelled (even if the cancellable does not yet exist). This is also covered by a unit test.